scripts/ansible/roles/avalanche_build/defaults/main.yml

```diff
@@ -1,5 +1,5 @@
-avalanche_binary: ~/go/src/github.com/ava-labs/avalanchego/build/avalanchego
-repo_folder: ~/go/src/github.com/ava-labs/avalanchego
+avalanche_binary: "{{ repo_folder }}/build/avalanchego"
+repo_folder: "~{{ ansible_facts.user_id }}/go/src/github.com/ava-labs/avalanchego"
```

Comment:
What's the purpose of using `ansible_facts.user_id` here? It looks like this variable is not defined when I'm attempting to run this role as part of the playbooks that use it.

Reply:
`ansible_facts` is a dictionary that is populated by Ansible. `ansible_facts.user_id` contains the username of the user that Ansible connected as (e.g. pi on a Raspberry Pi). I settled on `ansible_facts.user_id` as the best way to fix #291, following discussion with @holisticode.

Reply:
An example of its use
```
$ cat foo.yml
- hosts: localhost
  connection: local
  tasks:
    - debug:
       var: ansible_facts.user_id

$ ansible-playbook -i localhost, foo.yml 
[WARNING]: ansible.utils.display.initialize_locale has not been called, this may result in incorrectly
calculated text widths that can cause Display to print incorrect line lengths

PLAY [localhost] ****************************************************************************************

TASK [Gathering Facts] **********************************************************************************
ok: [localhost]

TASK [debug] ********************************************************************************************
ok: [localhost] => {
    "ansible_facts.user_id": "alex"
}

PLAY RECAP **********************************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0

Reply:
I see, seems like a good solution then. Will just have to add it to inventory files.

Reply:
Could you provide an example of how to do this in an inventory for one of the actual playbooks? I am trying to define this within an inventory for deploying to our devnet within the `vars` section, but am not sure how to do it based on this example

Reply:
To be clear, the playbook in my comment above (foo.yml) has nothing to do with Avalanche - it was just to illustrate an Ansible fact. At the moment you couldn't run a devnet or a testnet node with the `avalanche_service` role, I didn't add those options to the [avalanchego.service](https://github.com/ava-labs/avalanchego/blob/master/scripts/ansible/roles/avalanche_service/templates/avalanche.service) template. It's so far been used just on livenet, and everything added so far has been to support that.

I could add the remaining parameters if you like. If that were done, then your inventory would look something like 

https://github.com/moreati/avalanchego/blob/8dd9036fac484e6c3b1f67fd96a06bcd1bf64046/scripts/ansible/inventories/examples/amazon.yml#L1-L7

plus 

```
      network_id: ...
      bootstrap_ips: "{{ [hostvars[groups.avalanche_nodes[0]].ansible_facts.default_ipv4.address }}"
...
```

and your command would be

```
scripts/ansible/service_playbook.yml -i your_inventory.yml
```

Ansible also has inventory plugins - given the right credentials it can dynamically create an inventory based on EC2 API (or most cloud providers). I've not played much with that,

Hope that made sense, and that I didn't go off on the wrong tangent.

Reply:
That makes sense, my question is where would I define the value for `ansible_facts.user_id`. Reading the past conversation from #291 , I was under the impression that this should be defined in the inventory file.

Here is one inventory file that I was using (some ips changed):
```
devnet_bootstrap:
  hosts:
      bootstrap1:
        ansible_host: 127.127.127.127
        http_tls_enabled: false
        http_tls_key_file: ""
        http_tls_cert_file: ""
  vars:
    ansible_connection: ssh
    ansible_user: ubuntu
    user_id: ubuntu
    network_id: "local"
    api_admin_enabled: true
    api_keystore_enabled: true
    api_metrics_enabled: true
    tx_fee: 1000000
    assertions_enabled: true
    signature_verification_enabled: true
    db_enabled: true
    db_dir: "/home/ubuntu/db"
    http_port: 21000
    http_tls_enabled: false
    http_tls_key_file: ""
    http_tls_cert_file: ""
    http_host: ""
    bootstrap_ips: ""
    bootstrap_ids: ""
    staking_port: 21001
    staking_enabled: true
    staking_tls_key_file: "/home/ubuntu/go/src/github.com/ava-labs/avalanchego/staking/local/staker1.key"
    staking_tls_cert_file: "/home/ubuntu/go/src/github.com/ava-labs/avalanchego/staking/local/staker1.crt"
    plugin_dir: "/home/ubuntu/go/src/github.com/ava-labs/avalanchego/build/plugins"
    log_dir: "/home/ubuntu/.avalanchego"
    log_level: debug
    min_validator_stake: 10000000000
    min_delegator_stake: 5000000000
    staker_msg_reserved: 0.4
    staker_cpu_reserved: 0.4
    max_non_staker_pending_msgs: 20
    network_minimum_timeout: "500ms"
    network_initial_timeout: "10s"
    max_pending_msgs: 1024
    send_queue_size: 1024
    snow_sample_size: 3
    snow_quorum_size: 2
    snow_virtuous_commit_threshold: 20
    snow_rogue_commit_threshold: 30
    snow_avalanche_num_parents: 5
    snow_avalanche_batch_size: 30
    snow_epoch_first_transition: 1609347098
    snow_epoch_duration: "10s"
    api_ipcs_enabled: false
    xput_server_enabled: false
    xput_server_port: 21002
devnet_node:
  hosts:
      node1:
        ansible_host: 127.127.127.127
        staking_tls_key_file: "/home/ubuntu/go/src/github.com/ava-labs/avalanchego/staking/local/staker2.key"
        staking_tls_cert_file: "/home/ubuntu/go/src/github.com/ava-labs/avalanchego/staking/local/staker2.crt"
      node2:
        ansible_host: 127.127.127.127
        staking_tls_key_file: "/home/ubuntu/go/src/github.com/ava-labs/avalanchego/staking/local/staker3.key"
        staking_tls_cert_file: "/home/ubuntu/go/src/github.com/ava-labs/avalanchego/staking/local/staker3.crt"
      node3:
        ansible_host: 127.127.127.127
        staking_tls_key_file: "/home/ubuntu/go/src/github.com/ava-labs/avalanchego/staking/local/staker4.key"
        staking_tls_cert_file: "/home/ubuntu/go/src/github.com/ava-labs/avalanchego/staking/local/staker4.crt"
      node4:
        ansible_host: 127.127.127.127
        staking_tls_key_file: "/home/ubuntu/go/src/github.com/ava-labs/avalanchego/staking/local/staker5.key"
        staking_tls_cert_file: "/home/ubuntu/go/src/github.com/ava-labs/avalanchego/staking/local/staker5.crt"
  vars:
    ansible_connection: ssh
    ansible_user: ubuntu
    user_id: ubuntu
    network_id: "local"
    api_admin_enabled: true
    api_keystore_enabled: true
    api_metrics_enabled: true
    tx_fee: 1000000
    assertions_enabled: true
    signature_verification_enabled: true
    db_enabled: true
    db_dir: "/home/ubuntu/db"
    http_port: 21000
    http_tls_enabled: false
    http_tls_key_file: ""
    http_tls_cert_file: ""
    http_host: ""
    bootstrap_ips: "127.127.127.127:21001"
    bootstrap_ids: "NodeID-7Xhw2mDxuDS44j42TCB6U5579esbSt3Lg"
    staking_port: 21001
    staking_enabled: true
    plugin_dir: "/home/ubuntu/go/src/github.com/ava-labs/avalanchego/build/plugins"
    log_dir: "/home/ubuntu/.avalanchego"
    log_level: debug
    min_validator_stake: 10000000000
    min_delegator_stake: 5000000000
    staker_msg_reserved: 0.4
    staker_cpu_reserved: 0.4
    max_non_staker_pending_msgs: 20
    network_minimum_timeout: "500ms"
    network_initial_timeout: "10s"
    max_pending_msgs: 1024
    send_queue_size: 1024
    snow_sample_size: 3
    snow_quorum_size: 2
    snow_virtuous_commit_threshold: 20
    snow_rogue_commit_threshold: 30
    snow_avalanche_num_parents: 5
    snow_avalanche_batch_size: 30
    snow_epoch_first_transition: 1609347090
    snow_epoch_duration: "10s"
    api_ipcs_enabled: false
    xput_server_enabled: false
    xput_server_port: 21002
```
I tried defining the variable in this inventory file, but wasn't sure where that would go. When I'm running the update/restart playbooks, this variable is still saying that it is undefined, so my question is where in this inventory file would I define this so that those ansible scripts work properly?

Reply:
[`ansible_facts`](https://docs.ansible.com/ansible/latest/user_guide/playbooks_vars_facts.html#ansible-facts) is not a thing that you need to (or should) set. It is populated by Ansible itself when you run the play book. `ansible-playbook` prints "gathering facts when it does this. You can see the full list of facts available by running `ansible -i your_inventory.{ini,yml} -m setup` the output will look similar to this

```
localhost | SUCCESS => {
    "ansible_facts": {
        "ansible_all_ipv4_addresses": [
            "192.168.43.60",
            "172.17.0.1",
            "10.0.3.1",
            "192.168.122.1"
        ],
        "ansible_all_ipv6_addresses": [
            "2a01:4c8:a8:b609:fc91:19bc:e95d:8c69",
            "fe80::fa0f:85e:3155:96eb",
            "2a01:4c8:a8:b609:54a4:474c:23b1:6f49"
        ],
        "ansible_apparmor": {
            "status": "enabled"
        },
        "ansible_architecture": "x86_64",
        "ansible_bios_date": "02/05/2020",
...
```

Due to some historical cruft all the facts shown are prefixed `ansible_thing`, but the modern way to refer to them is `ansible_facts.thing`. Confusingly there are also "special variables" (that Ansible sets) and [connection variables](https://docs.ansible.com/ansible/latest/reference_appendices/special_variables.html#connection-variables) (that you can and should set in an inventory, or a playbook. Many of those (but not all) are *also* prefixed `ansible_`.

One connection variable is `ansible_user`. That's what the previous discussion was about, Since that discussion I switched from using `ansible_user` (a connection variable that you can set in an inventory), to `ansible_facts.user_id` (a measurement that is automatically performed by Ansible, on the target host(s)).

Hope that made sense, and wasn't too in the weeds.

Reply:
Worth mentioning: if you want to override repo_folder (or any other variable in these roles) you can ignore ansible_facts and connection variables entirely. Add something like this to the vars section of your inventory

```yaml
vars:
  repo_folder: //fo//bar/baz
 ```

PS sorry for the resolved/unresolved noise. I keep pressing the wrong button when adding a comment.

Reply:
I see, the problem here is that it is not actually being resolved and resulting in this error:
<img width="1468" alt="Screen Shot 2021-04-01 at 7 14 55 PM" src="https://user-images.githubusercontent.com/24684335/113363315-8d11fb80-931e-11eb-8704-9bbf63fb1df4.png">

So I think this needs to be resolved in order to work with our existing ansible scripts.

Reply:
Oh, it's because those playbooks have `gather_facts: false` . I'll remove that.

Reply:
https://github.com/ava-labs/avalanchego/pull/815/commits/c1ffc235f6150b3fe5e0fac765220064fea23b3d should fix it.

Sorry, I don't use any playbooks except service_playbook.yml, So I tried to leave as is, and not tread on your toes. Evidently that failed.

Reply:
No worries, thanks for resolving this. I think we could leave `gather_facts: false` for kill and ping playbooks since none of their roles use any of those variables.

When I'm running this now I'm seeing that the service playbook is failing due to a missing openssl dependency on the avalanche nodes when trying to create the SSL certs. It would be a good future change to make sure the dependencies are installed on the avalanche nodes, but I think that can be a part of a future PR, since so much great work has already gone into this one and it seems reasonable that a machine would have that dependency met (not to mention the prior version does not handle the SSL certs either).